temporian/implementation/numpy_cc/operators/calendar.cc

```diff
@@ -0,0 +1,258 @@
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+
+#include "absl/time/time.h"
+
+#include "temporian/implementation/numpy_cc/operators/common.h"
+
+namespace {
+namespace py = pybind11;
+
+static absl::TimeZone parse_tz(const py::object tz) {
+  if (py::isinstance<py::int_>(tz)) {
+    // handle tz as an int
+    const int int_tz = tz.cast<int>();
+    // tz is expresed in hours, needs to be converted to seconds
+    return absl::FixedTimeZone(int_tz * 60 * 60);
+
+  } else if (py::isinstance<py::float_>(tz)) {
+    // handle tz as an float
+    const float float_tz = tz.cast<float>() * 60 * 60;
+    const int int_tz = static_cast<int>(std::round(float_tz));
+    // tz is expresed in hours, needs to be converted to seconds
+    return absl::FixedTimeZone(int_tz);
+
+  } else if (py::isinstance<py::str>(tz)) {
+    // handle tz as an string
+    const std::string str_tz = tz.cast<std::string>();
+    absl::TimeZone parsed_tz;
+    if (!absl::LoadTimeZone(str_tz, &parsed_tz)) {
+      throw std::invalid_argument("Invalid timezone '" + str_tz +
+                                  "'. Only names defined in the IANA timezone" +
+                                  "database are valid");
+    }
+    return parsed_tz;
+
+  } else {
+    throw std::invalid_argument("Unsupported argument type for \"tz\" argument");
+  }
+}
+
+// TODO: support week starting on Sunday
+static int map_week_day(const absl::Weekday &wd) {
+  switch (wd) {
+    case absl::Weekday::monday:
+      return 0;
+    case absl::Weekday::tuesday:
+      return 1;
+    case absl::Weekday::wednesday:
+      return 2;
+    case absl::Weekday::thursday:
+      return 3;
+    case absl::Weekday::friday:
+      return 4;
+    case absl::Weekday::saturday:
+      return 5;
+    case absl::Weekday::sunday:
+      return 6;
+  }
+  std::invalid_argument("Invalid weekday");
+}
+
+// Template for iterating a timestamps array, converting to civil time
+// and apply a calendar_op to each one
+template <typename OUTPUT, typename TCALENDAR_OP>
+py::array_t<OUTPUT> apply_calendar_op(const py::array_t<double> &timestamps,
+                               const py::object tz) {
+  TCALENDAR_OP calendar_op;
+
+  auto parsed_tz = parse_tz(tz);
+
+  // Input size
+  const Idx n_events = timestamps.shape(0);
+
+  // Allocate output array
+  auto output = py::array_t<int>(n_events);
+
+  // Access raw input / output data
+  auto v_output = output.mutable_unchecked<1>();
+  auto v_timestamps = timestamps.unchecked<1>();
+
+  for (Idx i = 0; i < n_events; i++) {
+    // Create absolute time in UTC
+    auto nanos = static_cast<int64_t>(v_timestamps[i] * 1e9);
+    auto utc_time = absl::FromUnixNanos(nanos);
+
+    // Convert to civil time and get the isoweek
+    auto local_time = absl::ToCivilSecond(utc_time, parsed_tz);
+    v_output[i] = calendar_op.Get(local_time);
+  }
+
+  return output;
+}
+
+// Calendar operations interface
+template <typename OUTPUT> struct CALENDAR_OP {
```

Comment:
Since all the implementations return an "int", and none of them have any internal state, what about having the template parameter be simply a non templated function (instead of a templated class)?

Bonus: I don't think the speed gain of using a template is significant here (actually, it is likely to output the same code with the proper compilation flags). Maybe it would make the code simpler to use an argument such a: std::function<int(absl::Time)> function).

Reply:
You were right! I made the change and now there is much less boilerplate in the implementation